packages/navigator/src/browser/navigator-editor-sync.ts

```diff
@@ -0,0 +1,165 @@
+     *
+     * @param editor editor widget object to reveal its node in the navigator
+     */
+    selectNodeByEditor(editor: EditorWidget | undefined) {
+        if (editor) {
+            this.selectNodeById(this.editorIdToNavigatorNodeId(editor.id));
+        }
+    }
+
+    /**
+     * Converts editor id to navigator node id.
+     * Example: 'code-editor-opener:file:///home/user/workspace/README.md' => 'file:///home/user/workspace/README.md'
+     *
+     * @param editorId id of editor tab
+     * @returns id of corresponding navigator node
+     */
+    private editorIdToNavigatorNodeId(editorId: string) {
+        return editorId.substring(editorId.indexOf(':') + 1);
+    }
+
+    /**
+     * Reveals and selects node in the navigator by node id.
+     * If node with given id doesn't exist, nothing happens.
+     * Node id example: 'file:///home/user/workspace/README.md'
+     *
+     * @param nodeId navigator tree node id
+     */
+    selectNodeById(nodeId: string): void {
+        let node = this.fileNavigatorModel.getNode(nodeId);
+        if (node) {
+            // node is mounted in the navigator tree
+            if (ISelectableTreeNode.is(node)) {
+                this.revealNode(node);
+                this.fileNavigatorModel.selectNode(node);
+            }
+        } else {
+            // node may exist, but hasn't mounted yet
+            this.tryRevealNode(nodeId).then((targetNode: ITreeNode | undefined) => {
+                if (targetNode && ISelectableTreeNode.is(targetNode)) {
+                    // node exists and revealed in the navigator tree now
+                    this.fileNavigatorModel.selectNode(targetNode);
+                }
+            });
+        }
+    }
+
+    /**
+     * Reveals given node in navigator.
+     * This method should be used when given node already exist (but might be hidden) in the navigator tree.
+     * Otherwise use this.tryRevealNode
+     */
+    private revealNode(node: ITreeNode | undefined): void {
+        if (node) {
+            // cannot use ITreeNode.isVisible(node) here because it says that node is visible when it actually isn't
+            if (!('visible' in node) || node.visible === false) {
+                this.revealNode(node.parent);
+            }
+
+            if (IExpandableTreeNode.is(node) && !node.expanded) {
+                this.fileNavigatorModel.expandNode(node);
+            }
+        }
+    }
+
+    /**
+     * Tries to reveal node in the navigator by node id.
+     * Node id example: 'file:///home/user/workspace/src/subdir/file.ts'
+     *
+     * @param nodeId id of node to reveal
+     * @returns the node with given id if it exists, undefined otherwise
+     */
+    private async tryRevealNode(nodeId: string): Promise<ITreeNode | undefined> {
+        let rootNode = this.fileNavigatorModel.root;
+        if (rootNode && nodeId.startsWith(rootNode.id)) {
+            let segments = nodeId.substring(rootNode.id.length + 1).split('/');
```

Comment:
We are in the navigator, so you can assume, the `id` is a URI string. Please create a `new URI(id)`, and then you can have `uri.parent` to traverse up the hierarchy.

Reply:
Thanks, great idea